File: apps/iosApp/Sources/ViewModels/BrowseViewModel.swift
import Foundation
import SharedKit

@MainActor
final class BrowseViewModel: ObservableObject {
    @Published private(set) var items: [ContentSummary] = []
    @Published var selectedType: ContentType?
    @Published private(set) var isLoading = true
    @Published private(set) var isLoadingMore = false
    @Published private(set) var hasMore = false
    @Published private(set) var error: String?

    private var cursor: String?
    private let contentRepository: ContentRepository

    init(contentRepository: ContentRepository) {
        self.contentRepository = contentRepository
        load()
    }

    func selectType(_ type: ContentType?) {
        selectedType = type
        items = []
        cursor = nil
        hasMore = false
        load()
    }

    func load() {
        Task {
            isLoading = true
            error = nil
            do {
                let query = ContentQuery(type: selectedType, genre: nil, releaseYear: nil, cursor: nil, limit: 24)
                let page = try await ThrowingWrappersKt.listOrThrow(contentRepository, query: query)
                items = (page.items as! [ContentSummary])
                hasMore = page.hasMore
                cursor = page.nextCursor
                isLoading = false
            } catch {
                self.error = readableMessage(error)
                isLoading = false
            }
        }
    }

    func loadMore() {
        guard let cursor, !isLoadingMore, hasMore else { return }
        Task {
            isLoadingMore = true
            do {
                let query = ContentQuery(type: selectedType, genre: nil, releaseYear: nil, cursor: cursor, limit: 24)
                let page = try await ThrowingWrappersKt.listOrThrow(contentRepository, query: query)
                items += (page.items as! [ContentSummary])
                hasMore = page.hasMore
                self.cursor = page.nextCursor
                isLoadingMore = false
            } catch {
                isLoadingMore = false
            }
        }
    }
}

File: apps/androidApp/src/main/kotlin/com/streaming/platform/android/ui/BrowseViewModel.kt
package com.streaming.platform.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.streaming.platform.content.ContentQuery
import com.streaming.platform.content.ContentRepository
import com.streaming.platform.content.ContentSummary
import com.streaming.platform.content.ContentType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class BrowseUiState(
    val items: List<ContentSummary> = emptyList(),
    val selectedType: ContentType? = null,
    val isLoading: Boolean = true,
    val isLoadingMore: Boolean = false,
    val hasMore: Boolean = false,
    val error: String? = null,
    private val cursor: String? = null,
) {
    val nextCursor: String? get() = cursor
}

class BrowseViewModel(private val contentRepository: ContentRepository) : ViewModel() {
    private val _uiState = MutableStateFlow(BrowseUiState())
    val uiState: StateFlow<BrowseUiState> = _uiState.asStateFlow()

    init { load() }

    fun selectType(type: ContentType?) {
        _uiState.value = BrowseUiState(selectedType = type)
        load()
    }

    fun load() {
        val state = _uiState.value
        viewModelScope.launch {
            _uiState.value = state.copy(isLoading = true, error = null)
            contentRepository.list(ContentQuery(type = state.selectedType, limit = 24)).fold(
                onSuccess = { page ->
                    _uiState.value = BrowseUiState(
                        items = page.items,
                        selectedType = state.selectedType,
                        isLoading = false,
                        hasMore = page.hasMore,
                        cursor = page.nextCursor,
                    )
                },
                onFailure = { _uiState.value = state.copy(isLoading = false, error = it.readableMessage()) },
            )
        }
    }

    fun loadMore() {
        val state = _uiState.value
        val cursor = state.nextCursor ?: return
        if (state.isLoadingMore || !state.hasMore) return
        viewModelScope.launch {
            _uiState.value = state.copy(isLoadingMore = true)
            contentRepository.list(ContentQuery(type = state.selectedType, cursor = cursor, limit = 24)).fold(
                onSuccess = { page ->
                    _uiState.value = state.copy(
                        items = state.items + page.items,
                        isLoadingMore = false,
                        hasMore = page.hasMore,
                        cursor = page.nextCursor,
                    )
                },
                onFailure = { _uiState.value = state.copy(isLoadingMore = false, error = it.readableMessage()) },
            )
        }
    }
}
